sql/expression/function/logarithm.go

```diff
@@ -0,0 +1,223 @@
+package function
+
+import (
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"fmt"
+	"gopkg.in/src-d/go-errors.v1"
+	"math"
+	"reflect"
+)
+
+// ErrInvalidArgumentForLogarithm is returned when an invalid argument value is passed to a
+// logarithm function
+var ErrInvalidArgumentForLogarithm = errors.NewKind("invalid argument value for logarithm: %v")
+
+// LogBaseMaker returns LogBase creator functions with a specific base.
+func LogBaseMaker(base float64) func(e sql.Expression) sql.Expression {
+	b := base
+	return func(e sql.Expression) sql.Expression {
+		return NewLogBase(b, e)
+	}
+}
+
+// LogBase is a function that returns the logarithm of a value with a specific base.
+type LogBase struct {
+	expression.UnaryExpression
+	base float64
+}
+
+// NewLogBase creates a new LogBase expression.
+func NewLogBase(base float64, e sql.Expression) sql.Expression {
+	return &LogBase{UnaryExpression: expression.UnaryExpression{Child: e}, base: base}
+}
+
+func (l *LogBase) String() string {
+	switch l.base {
+	case float64(math.E):
+		return fmt.Sprintf("ln(%s)", l.Child)
+	case float64(10):
+		return fmt.Sprintf("log10(%s)", l.Child)
+	case float64(2):
+		return fmt.Sprintf("log2(%s)", l.Child)
+	default:
+		return fmt.Sprintf("log(%v, %s)", l.base, l.Child)
+	}
+}
+
+// TransformUp implements the Expression interface.
+func (l *LogBase) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := l.Child.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	return f(NewLogBase(l.base, child))
+}
+
+// Type returns the resultant type of the function.
+func (l *LogBase) Type() sql.Type {
+	return sql.Float64
+}
+
+// IsNullable implements the sql.Expression interface.
+func (l *LogBase) IsNullable() bool {
+	return l.Child.IsNullable()
+}
+
+// Eval implements the Expression interface.
+func (l *LogBase) Eval(
+	ctx *sql.Context,
+	row sql.Row,
+) (interface{}, error) {
+	v, err := l.Child.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if v == nil {
+		return nil, nil
+	}
+
+	val, err := sql.Float64.Convert(v)
+	if err != nil {
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(v))
+	}
+	return computeLog(val.(float64), l.base)
+}
+
+// Log is a function that returns the natural logarithm of a value.
+type Log struct {
+	expression.BinaryExpression
+}
+
+// NewLn creates a new Log expression.
+func NewLog(args ...sql.Expression) (sql.Expression, error) {
+	argLen := len(args)
+	if argLen == 0 || argLen > 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New("1 or 2", argLen)
+	}
+
+	var right sql.Expression
+	if len(args) == 2 {
+		right = args[1]
+	}
+
+	return &Log{expression.BinaryExpression{Left: args[0], Right: right}}, nil
```

Comment:
isn't better to always set `b` and `x`  instead of checking somewhere else which one is nil?
- if len == 1  => left = E, right = args[0]
- if len == 2 => left = args[0], right = args[1]

Reply:
OK will refactor but I took this logic from 
https://github.com/src-d/go-mysql-server/blob/master/sql/expression/function/ceil_round_floor.go#L155-L166

Reply:
I believe, internally it's better to have everything explicit.